exercises/generic_sorting/descr.md

```diff
@@ -0,0 +1,299 @@
+**description** of a total preorder on values of type `'k`.
+For every such value `o`, we are able to give a linear-time sorting algorithm.
+This claim is substantiated by defining a function `sort` whose type is as follows:
+
+```
+  val sort: 'k order -> ('k * 'v) list -> 'v list
+```
+
+The function `sort o` accepts a list of key-value pairs,
+sorts it according to the preorder `o`,
+which is a preorder on keys,
+and returns just a list of values.
+The keys are dropped.
+(This somewhat unusual presentation, where keys and values are distinguished,
+helps define `sort` recursively. A more standard presentation, where keys and
+values are identified, can easily be obtained as a special case; see Question 4.)
+
+The time complexity of `sort o kvs` is `O(n)`, where `n` is the length of the
+input list `kvs`, and where the constant factor is allowed to depend on the
+preorder `o`.
+
+Throughout this exercise, we restrict our attention to **stable** sorting
+algorithms. That is, if two values are associated with equivalent keys
+according to the preorder `o`, then they must appear in the output list in the
+same order as in the input list.
+
+This exercise is inspired by the paper
+[Generic top-down discrimination for sorting and partitioning in linear
+time](https://www.cs.ox.ac.uk/projects/utgp/school/henglein2012c.pdf)
+(Henglein, 2012).
+A related paper is
+[Sorting and searching by distribution: from generic discrimination to
+generic tries](http://www.cs.ox.ac.uk/ralf.hinze/WG2.8/31/slides/aplas2013.pdf)
+(Henglein and Hinze, 2013).
+
+The algorithms considered here and in these papers are *generic*. Here, this
+word means that these algorithms are polymorphic and directed by an explicit
+description of the data at hand. Indeed, they can be applied to an arbitrary
+type `'k` of keys, provided a description `o` of a preorder on keys, of type
+`'k order`, is provided.
+
+## Warmup: Pigeonhole Sort
+
+**Question 1.** Define a function `pigeonhole_sort`
+of type `int -> (int * 'v) list -> 'v list`
+such that `pigeonhole_sort bound kvs`
+sorts the list of key-value pairs `kvs`
+and returns just a list of values.
+The keys are integers
+and must be comprised between `0` (included) and `bound` (excluded).
+The time complexity of the algorithm must be `O(bound + n)`,
+where `n` is the length of the list `kvs`.
+This must be a stable sort.
+
+*Hint.* Build an array, indexed by keys,
+where each array cell holds a list of values.
+Convert this array back to a sorted list.
+
+*Note.* The automatic grading code verifies whether the output list is
+correct, but does not verify the time complexity of your algorithm. It
+is up to you to convince yourself that the complexity bound is met.
+
+*Note.* In a real-world implementation,
+one would use memoization so as to ensure
+that two successive calls to `pigeonhole_sort bound`
+use the same internal array,
+instead of allocating a new array every time.
```

Comment:
That's not clear to me as this would prevent concurrent executions of this function.

Reply:
OK, reformulated in a less affirmative way.